linera-views/src/reentrant_collection_view.rs

```diff
@@ -272,6 +272,21 @@ where
     pub async fn try_load_entry(
         &self,
         short_key: Vec<u8>,
+    ) -> Result<ReadGuardedView<W>, ViewError> {
+        {
+            let mut hash = self.hash.lock().await;
+            *hash = None;
+        }
+        self.local_try_load_entry(short_key).await
```

Comment:
The fact that we need to invalidate the hash for `&self` function is a bit concerning because `&self` is how we restrict views to be read-only. Have you looked at making `try_load_entry` failed instead if the entry doesn't exist?

Reply:
That is concerning, and I created the issue #1442 to discuss it in more detail.

The problem is that with the storage code, existing on the storage is the same as not being default. For a `ReentrantCollectionView<CollectionView>` that would mean a try_load_entry would fail if the collection is empty, which would be a pretty radical step.

Reply:
@MathieuDutSik Let's go directly for the correct design on all counts.
* In the rest of the `linera-view`, indeed, a view always "exists" in its default state, or to say it differently, `load` doesn't create new objects.
* Note that `fn get(&self)` doesn't create a default element in a `MapView`. This would be wrong because the set of elements is observable (elements are both iterable and hashable). In fact, we have correctly distinguished `get_mut` and `get_mut_or_default`.
* On the other hand, in `*CollectionView`s, we have used the concept of `load_entry` with the meaning of `load_entry_or_insert`. The difference matters because the set of entries in a collection *is also observable*. Because of the confusing name, we have allowed `&self` methods to mutate the list of entries but that is incorrect. I don't think we rely on this (wrong) behavior in the rest of the code. So let's introduce the distinction between `load_entry_or_insert` (requires `&mut self`) and `*load_entry` (cannot mutate anything, doesn't require hash invalidation).

Interestingly, we don't have the problem with `linera_storage::Storage::load_chain` because the list of all chains ever loaded is not observable. (This plus the fact that hashing was added later might explain why we had the wrong idea for collection views.)

Reply:
Thank you for going directly to this. Indeed it works fine and I could implement the design that you proposed.